### PR TITLE
[JENKINS-66105]The appropriate HTTP code in buildWithParameters.

### PR DESCRIPTION
**Pull request by i-am-vt**
In buildWithParameters api, when a new parameter build request is determined to be repeated, HTTP 201 should not be returned here. 409 is generally considered to be more appropriate. But since Jenkins actually merged some foldable actions, I think we should return 303 here.

image
As shown in the pic above, it is normal for the request of #15 to response http code 201, however, as long as # 15 is still in the queue, no matter how many requests with the same parameters as #15 will be merged, but api still response 201 http code,
this makes it impossible for clients to tell which requests will actually be built.

See [JENKINS-66105](https://issues.jenkins-ci.org/browse/JENKINS-66105).

### Proposed changelog entries

* Entry 1: Issue, When the **buildWithParameter** API is called, if the requests with the same parameters in the queue are merged, the http response code of the request uses a more appropriate 303(see other) instead of 201(created).

### Proposed upgrade guidelines

==== Optimize the http response code of buildWithParameters api

[https://issues.jenkins-ci.org/browse/JENKINS-66105](https://issues.jenkins-ci.org/browse/JENKINS-66105)
In the previous version, when calling the buildWithParameters rest api `{jenkins_root_path}/job/{job_name}/buildWithParameters` of Jenkins, if Jenkins judges that the queue contains requests with the same parameters and merges them, the merged request would also return [HTTP 201](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201), making the client unable to determine which requests were built. 
To solve this problem, in the current version, if the same params request in the queue is merged, [HTTP 303](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303) will be returned, and this is more in line with http semantics.

For example, a job contains a parameter named 'test-key'.
Execute the following http request command:
```shell
curl -XPOST -i -s -d 'test-key=1' 'http://{jenkins_host}/job/{job_name}/buildWithParameters?delay=10sec' | grep 'HTTP\|Location'
echo '------'  
curl -XPOST -i -s -d 'test-key=1' 'http://{jenkins_host}/job/{job_name}/buildWithParameters?delay=10sec' | grep 'HTTP\|Location'
```
Output before the change is rough as follows:
```shell
HTTP/1.1 201
Location: http://{jenkins_host}/queue/item/26/
------
HTTP/1.1 201
Location: http://{jenkins_host}/queue/item/26/
```
After the change:
```shell
HTTP/1.1 201 
Location: http://{jenkins_host}/queue/item/27/
------
HTTP/1.1 303
Location: http://{jenkins_host}/queue/item/27/
```
We can see that the request first created after the change still returns http 201 unaffected, and the http code of merged request return becomes 303.

**Note:** that if you or the request tool you are using has checked the http response code, you must confirm if the new scenario returning 303 affects you, otherwise this could cause some failure.

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
> Entry 1: Issue, The appropriate HTTP code in buildWithParameters.

This does not look like it means much to the general changelog audience. Please think of more appropriate user documentation for this change.

**Comment by i-am-vt** [comment]
> > Entry 1: Issue, The appropriate HTTP code in buildWithParameters.
> 
> This does not look like it means much to the general changelog audience. Please think of more appropriate user documentation for this change.

Following your advice, I have updated some contents:)

**Review by oleg-nenashev — APPROVED** [review]
No objections from me. Would be nice to get more reviews from the core team though

**Review by jtnord — APPROVED** [review]
Code looks ok, I wonder what if any external scripts / tooling this may break and if this should be given a breaking status (certainly clients that where expecting a 2xx response will now fail and get a 3xx)

**Comment by oleg-nenashev** [comment]
I agree with @jtnord, it would be nice to notify users. And upgrade guide entry would be nice. @i-am-vt would you be available to write one?

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by i-am-vt** [comment]
> I agree with @jtnord, it would be nice to notify users. And upgrade guide entry would be nice. @i-am-vt would you be available to write one?


Of course, I just updated the upgrade guide.

**Comment by oleg-nenashev** [comment]
Thanks @i-am-vt !

**Comment by jonsie** [comment]
Heads up we are seeing problems associated to this change.  In every instance, when following the 303 we are hitting a 404 instead of actually retrieving the `queue/item`.  Will do a bit more testing, but seems like a potential bug.

**Comment by jglick** [comment]
> when following the 303 we are hitting a 404 instead of actually retrieving the `queue/item`.

You cannot follow the `/queue/item/$number/` link directly (for example, `curl -L` will not work) because this URL root does not have content. Instead you would append something to the `Location`, such as `api/json?tree=task[url]`.

**Pull request merged**
